autotools: add attribute to delete libtool archives .la files

alalazo opened the pull request:
fixes #18694

According to [Autotools Mythbuster](https://autotools.io/libtool/lafiles.html) libtool archive files are mostly vestigial, but they might create issues when relocating binary packages as shown in #18694. For GCC specifically, most distributions remove these files [with explicit commands](https://git.stg.centos.org/rpms/gcc/blob/master/f/gcc.spec#_1303).

~Considered all of that, this commit adds an easy way for each `AutotoolsPackage` to remove every .la file that has been installed. The default, for the time being, is to maintain them - to be consistent with what Spack was doing previously.~

This PR makes removing `.la` files for all autotools packages the default behavior, and leaves the option to revert to the old behavior on a package by package base.

alalazo committed:
autotools: add attribute to delete libtool archives .la files

fixes #18694

According to Autotools Mythbuster (https://autotools.io/libtool/lafiles.html)
libtool archive files are mostly vestigial, but they might create issues
when relocating binary packages as shown in #18694.

For GCC specifically, most distributions remove these files with
explicit commands:

https://git.stg.centos.org/rpms/gcc/blob/master/f/gcc.spec#_1303

Considered all of that, this commit adds an easy way for each
AutotoolsPackage to remove every .la file that has been installed.
The default, for the time being, is to maintain them - to be consistent
with what Spack was doing previously.

alalazo commented:
@eugeneswalker Can you check if this works for you?

eugeneswalker commented:
> @eugeneswalker Can you check if this works for you?

Yes, this worked for me!

michaelkuhn reviewed (approved):
Just gave this a quick test with a (modified) freetype package and it worked fine. I also see no harm, since Linux distributions have been doing this for years.

eugeneswalker commented:
In a general sense I understand the tact of preserving existing behavior, but should we consider doing it the other way, i.e. making it so that this is ON by default?  If these libtool archive files are in fact vestigial, and if in fact many linux distros already do this, perhaps the better way forward is to break existing behavior, since the existing behavior is in a sense *already broken* as described in the linked issue. Any thoughts? If we proceed with this PR as is, is there any alternative than to extensively seek out every package that is going to have this issue? I don't think that will be a trivial task, since, as was the case with mpich, affected packages will *appear to install just fine from build cache* and only show problems when run.

becker33 reviewed (changes requested):
I think this needs a test, and I think @eugeneswalker is right that we should change the default and provide an option to preserve behavior.

alalazo commented:
> In a general sense I understand the tact of preserving existing behavior, but should we consider doing it the other way, i.e. making it so that this is ON by default?

I think doing a breaking change is the best option, since there's agreement on that. I'll modify the PR and add a test by today.

alalazo committed:
autotools: delete libtool archive files by default

Following review this commit changes the default for
libtool archive files deletion and adds test to verify
the behavior.

alalazo commented:
> If we proceed with this PR as is, is there any alternative than to extensively seek out every package that is going to have this issue? I don't think that will be a trivial task, since, as was the case with mpich, affected packages will appear to install just fine from build cache and only show problems when run.

There's a note  I'd like to make explicitly. While I agree on the new defaults, the issue in #18694 was triggered by the libtool archive files being part of the compiler. Once you remove them **only** for the compilers (which are currently treated differently from all other dependencies) everything should work since we have already a mechanism to do path substitution on text files during relocation of "normal" dependencies. 

This is just to say that I wouldn't expect a lot more packages than just GCC to be affected by this - but I indeed like the idea to remove `.la` files all together by default for the reasons already said and also because their intended purpose is to my understanding overlapping with our compiler wrapper.

alalazo commented:
@becker33 @eugeneswalker This is ready for a further review.

eugeneswalker commented:
This looks good to me! Thank you for doing this!

michaelkuhn reviewed (approved):
I also did a quick test with the new version: still works for me.

tjahns commented:
As someone who develops software that uses libtool I find this proposal to be misguided:
1. It becomes harder to use spack-deployed software this way (libtool no longer picks up the -rpath argument automatically)
2. .la-files are only one component of packages that contains the install directory, pkg-config files and the binaries themselves (especially when the package contains plugins) are examples that are similarly relevant.
3. Distributions can drop .la files because their libraries go into directories searched by ld and ld.so by default.
4. Why was this feature not introduced with a global off-switch?
5. Since I just had to explicitly add -Wl,-rpath arguments to link lines in multiple builds because of this "feature", I find the characterization of .la files as vestigial quite incomplete.
